File: data structures and algorithms/java/08_068.java
class PlayerNode {
   String name;
   double battingAverage;
   PlayerNode next;
   
   public PlayerNode(String name, 
         double battingAverage) {
      this.name = name;
      this.battingAverage = battingAverage;
      this.next = null;
   }
}

public class Solution {
   public String rotateLineupAndFindBest(
         PlayerNode head, int k, int m) {
      // Handle edge cases
      if (head == null || head.next == head || 
          k == 0) {
         return findBestBatter(head, m);
      }
      
      // Find kth node
      PlayerNode current = head;
      for (int i = 0; i < k - 1; i++) {
         current = current.next;
      }
      
      // New head
      PlayerNode newHead = current.next;
      
      // Find last node
      PlayerNode last = newHead;
      while (last.next != head) {
         last = last.next;
      }
      
      // Rotate circular list
      current.next = head;
      last.next = newHead;
      
      // Find best batter
      return findBestBatter(newHead, m);
   }
   
   // Find best batter in window
   private String findBestBatter(PlayerNode head, 
         int m) {
      if (head == null) {
         return null;
      }
      
      PlayerNode bestPlayer = head;
      PlayerNode current = head;
      
      // Check first m players
      for (int i = 0; i < m - 1; i++) {
         current = current.next;
         if (current.battingAverage > 
             bestPlayer.battingAverage) {
            bestPlayer = current;
         }
         
         // Handle circular boundary
         if (current == head) {
            break;
         }
      }
      
      return bestPlayer.name;
   }
}

File: data structures and algorithms/cs/08_068.cs
using System;

class PlayerNode
{
   public string Name { get; }
   public double BattingAverage { get; }
   public PlayerNode Next { get; set; }

   public PlayerNode(
      string name,
      double battingAverage)
   {
      Name = name;
      BattingAverage = battingAverage;
      Next = null;
   }
}

public class Solution
{
   public string RotateLineupAndFindBest(
      PlayerNode head,
      int k,
      int m)
   {
      if (head == null || head.Next == head || k == 0)
      {
         return FindBestBatter(head, m);
      }

      PlayerNode current = head;
      for (int i = 0; i < k - 1; i++)
      {
         current = current.Next;
      }

      PlayerNode newHead = current.Next;

      PlayerNode last = newHead;
      while (last.Next != head)
      {
         last = last.Next;
      }

      current.Next = head;
      last.Next = newHead;

      return FindBestBatter(newHead, m);
   }

   private string FindBestBatter(
      PlayerNode head,
      int m)
   {
      if (head == null)
      {
         return null;
      }

      PlayerNode bestPlayer = head;
      PlayerNode current = head;

      for (int i = 0; i < m - 1; i++)
      {
         current = current.Next;
         if (current.BattingAverage >
            bestPlayer.BattingAverage)
         {
            bestPlayer = current;
         }

         if (current == head)
         {
            break;
         }
      }

      return bestPlayer.Name;
   }
}
